src/constraints.jl

```diff
@@ -727,3 +727,313 @@ function list_of_constraint_types(model::Model)
     return Tuple{DataType, DataType}[(jump_function_type(model, f), s)
                                      for (f,s) in list]
 end
+
+
+"""
+    perturbation_range_of_feasibility(constraint::ConstraintRef)
+
+Gives the range by which the rhs coefficient can change and the current lp-basis
+remains feasible, i.e., where the shadow prices apply.
+
+## Notes
+- The range denote valid changes, e.g., for a*x <= b + Δ, the lp-basis remains feasible for all Δ in [l, u].
+
+"""
+function perturbation_range_of_feasibility(constraint::ConstraintRef{Model, <:_MOICON})
+    error("The range of feasibility is not defined or not implemented for this type " *
+          "of constraint.")
+end
+
+"""
+Builds the standard form constraint matrix, varaible bounds (rhs := 0, by choice of slack bounds), and a vector of constraint reference one for each row.
+i.e., the LP-problem feasible set on the form
+s.t.  Ax == 0
+L <= x <= U
+"""
+function _std_matrix(model::Model)
+    con_types = list_of_constraint_types(model)
+    con_func_type = first.(con_types)
+    if !all(broadcast(<:, AffExpr, con_func_type) .| broadcast(<:, VariableRef, con_func_type))
+        error("Range of feasibility and optimality require all constraints functions to be affine (the problem to be linear).")
+    end
+    vars = all_variables(model)
+    var2idx = Dict{VariableRef, Int64}()
+    for (j, var) in pairs(vars)
+        var2idx[var] = j
+    end
+    col_j = Int64[]
+    row_i = Int64[]
+    coeffs = Float64[]
+
+    var_lower = fill(-Inf, length(vars))
+    var_upper = fill(Inf, length(vars))
+    affine_constraints = ConstraintRef[]
+
+    m = 0
+    n = length(vars)
+    for (F, S) in con_types
+        constraints = all_constraints(model, F, S)
+        if F <: VariableRef
+            for (i, constr) in pairs(constraints)
+                constr_obj = constraint_object(constr)
+                j = var2idx[constr_obj.func]
+                if S <: MOI.EqualTo
+                    var_lower[j] = var_upper[j] = constr_obj.set.value
+                end
+                if S <: MOI.Interval
+                    var_lower[j] = constr_obj.set.lower
+                    var_upper[j] = constr_obj.set.upper
+                end
+                if S <: MOI.LessThan
+                    var_upper[j] = constr_obj.set.upper
+                end
+                if S <: MOI.GreaterThan
+                    var_lower[j] = constr_obj.set.lower
+                end
```

Comment:
Error if the set isn't one of the 4 types checked for. This code will silently ignore integrality constraints.